### DISCUSSION TITLE
Can we use "squash" merging instead of "merge commit" for Pull Requests?

### DISCUSSION THREAD

**Discussion by wookayin** [jax-ml/jax#13265] [General]
Can we have pull requests from OSS contributors merged by "squashing" or some kind of "rebasing" instead of the default strategy "merging" (i.e., creating a merge commit)?

The reason why I'm suggesting this is to make git graph look much nicer and does not look intimidating?

<img width="1297" alt="image" src="https://user-images.githubusercontent.com/1009873/202048386-aa01c8f7-9705-4a46-b12e-f85e8b904d49.png">

<img width="1297" alt="image" src="https://discuss.tensorflow.org/uploads/default/original/2X/6/68128889b7389dc267e474f83fffa2746704400e.png" />

In my opinion, such octopus-like merge commits ("Merge Pull Request ...") are basically useless and has no practical meaning. It would be really great if the git history could be maintained linearly and clean, so that non-Googlers & OSS contributors can better read the commit history. The non-linear commit graph makes searching and tracking of history **really difficult**, given the large size of the repository.

I know Google uses their internal VCS (Piper) tool and use Copybara to make synchronization between Github and Piper. According to @mihaimaruseac's [comment](https://discuss.tensorflow.org/t/suggestion-for-oss-pr-workflow-rebasing-instead-of-merging/5700/3), every (internal) changelist in Piper associated with a PR must be atomic; **each PR must correspond to just one single CL**. However, with Github's PR workflow we can do this with the "squash-and-merge" strategy. This should be possible with Copybara and [Github's repo configuration](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests), and I strongly suggest jax (and other projects) should use this merging strategy by default so as to be more open-source friendly.

I have [asked the same thing for TensorFlow](https://discuss.tensorflow.org/t/suggestion-for-oss-pr-workflow-rebasing-instead-of-merging/5700) (as well as many other Google's OSS projects), but it hasn't been responded or gained much attention for many years.

**Comment by froystig** [comment]
We too have said at several points that standard github-style squash-merges would be a nice improvement. Maybe we should move this discussion to the [Copybara project's github](https://github.com/google/copybara) to track it there?

**Reply by wookayin** [reply]
Thank you @froystig for the response. I see, so it looks like that at the moment Copybara doesn't have such features implemented yet? I'm fine with moving the discussion there and having the Copybara team more involved together.

**Reply by froystig** [reply]
My understanding is that Copybara handles the merge by pushing a git commit (i.e. an explicit merge node) rather than via Github API queries. Either way, seems good to raise with them.

**Comment by wangyendt** [comment]
This is an auto-sent e-mail. 
 
I am sorry that currently I could not reply your e-mail. I will reply you as soon as possible.

**Comment by LONGMANKE** [comment]
Such "Merge Pull Request..." merge commits that resemble an octopus, in my opinion, are essentially pointless and have no real world application. It would be fantastic if the git history could be kept neat and linear so that OSS contributors and anyone outside of Google could more easily read the commit history. Given the scale of the repository, the non-linear commit graph makes searching and tracking history quite challenging.

I am aware that Google synchronizes Github and Piper via Copybara and their internal VCS tool, Piper. According to @mihaimaruseac's reply, each Piper PR must correspond to a single CL, and every (internal) changelist connected with a PR must be atomic.

**Comment by wangyendt** [comment]
This is an auto-sent e-mail. 
 
I am sorry that currently I could not reply your e-mail. I will reply you as soon as possible.
